### PR TITLE
AirfRANS: Huber loss on volume channel for robust optimization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The volume MSE loss is sensitive to outlier gradient magnitudes — a small fraction of cells with large prediction errors dominate the gradient signal and destabilize volume optimization. Replacing the volume channel's MSE loss with **Huber loss (SmoothL1)** should suppress these outlier gradients: for errors smaller than delta, Huber behaves like MSE (quadratic); for larger errors it falls back to L1 (linear). This reduces the variance of the gradient signal coming from volume errors without completely abandoning the strong quadratic pull near zero.

The champion config already achieves a programme-best vol_mse = 0.002039 with `--vol-loss-weight 10x`. We believe a tighter vol_mse (target: < 0.0017) is achievable by making the volume loss more robust, without sacrificing the current surface_mse champion of 0.000296.

Reference: Huber (1964), also implemented as `torch.nn.SmoothL1Loss(beta=delta)` or `torch.nn.HuberLoss(delta=delta)` in PyTorch.

## Instructions

**Hard constraint:** surface_mse must NOT regress below 0.000296. If any trial causes surface_mse > 0.000296, note it but do not treat it as a winner.

**Target:** vol_mse < 0.0017 (current champion: 0.002039).

Use `--wandb_group af-huber-volume` for all runs so related trials are grouped in W&B.

### Implementation

In `target/icml2026/train.py`, locate the volume loss computation. Replace the MSE volume loss with Huber loss:

```python
# Before (MSE):
vol_loss = F.mse_loss(vol_pred, vol_target)

# After (Huber / SmoothL1):
vol_loss = F.huber_loss(vol_pred, vol_target, delta=DELTA)
# OR equivalently:
# vol_loss = F.smooth_l1_loss(vol_pred, vol_target, beta=DELTA)
```

Keep the surface loss as MSE (unchanged). Only the volume channel loss changes.

Add a `--vol-loss-delta` argument (float, default=1.0) to `train.py` to control the Huber delta at runtime so no code changes are needed between trials.

### Smoke test (run first before full trials)

Run 3 epochs to confirm no crashes and that vol_loss is well-behaved:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --ema-decay 0.999 --vol-loss-weight 10.0 \
  --vol-loss-delta 1.0 \
  --wandb_group af-huber-volume \
  --epochs 3
```

### Trial 1 — Huber delta=1.0 (full run)

Full run with the champion config and Huber delta=1.0:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --ema-decay 0.999 --vol-loss-weight 10.0 \
  --vol-loss-delta 1.0 \
  --wandb_group af-huber-volume \
  --epochs 999
```

### Trial 2 — Huber delta=0.1 (full run)

A much tighter transition threshold — forces L1 behavior for a wider range of errors. This is more aggressive outlier suppression:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --ema-decay 0.999 --vol-loss-weight 10.0 \
  --vol-loss-delta 0.1 \
  --wandb_group af-huber-volume \
  --epochs 999
```

### Trial 3 — Best delta + vol-weight=15x (conditional)

**Only run Trial 3 if both Trial 1 and Trial 2 improve vol_mse below 0.002039.**

Take whichever delta value performed better (1.0 or 0.1) and increase the volume loss weight to 15x:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --ema-decay 0.999 --vol-loss-weight 15.0 \
  --vol-loss-delta <BEST_DELTA_FROM_T1_OR_T2> \
  --wandb_group af-huber-volume \
  --epochs 999
```

### Reporting

Report surface_mse and vol_mse from both the best-val-surface-MSE checkpoint and the best-val-vol-MSE checkpoint for each trial. Include W&B run IDs.

## Baseline

Current AirfRANS champion (PR #3135):

| Metric | Value | Epoch |
|--------|-------|-------|
| val_primary/surface_mse | **0.000296** | 704 |
| val_primary/vol_mse | **0.002039** | 743 |

- W&B baseline run: `sh2zyfwr` (nezuko/af-ema-vol-weight-10x)
- PR: #3135
- Config: 2L/256d/4H, AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, EMA=0.999, vol-weight=10x, Fourier

**Reproduce command:**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --ema-decay 0.999 --vol-loss-weight 10.0
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### Summary

**Hypothesis FALSIFIED.** Huber loss on the volume channel does not improve on the MSE baseline at either delta tested. Neither trial beats the baseline on surface_mse or vol_mse.

### Baseline (PR #3135, W&B: `sh2zyfwr`)

| Metric | Baseline |
|--------|----------|
| val surface_mse | 0.000296 (ep704) |
| val volume_mse | 0.002039 (ep743) |

### Trial 1 — Huber delta=1.0 (W&B: `87hwn7by`)

| Metric | Value | vs Baseline |
|--------|-------|-------------|
| **val surface_mse** | 0.000484 (ep327) | **+63.5%** ❌ |
| val volume_mse (at best srf) | 0.005460 | +168% ❌ |
| test surface_mse | 0.002018 | — |
| test volume_mse | 0.006784 | — |
| Terminal val vol_mse (ep762) | 0.002874 | +41% ❌ |
| Total epochs | 762 (360-min timeout) | |
| Grad norm trend | 22→6 (stable, no divergence) | |
| Peak VRAM | ~24 GB | |

**Command:**
```
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --ema-decay 0.999 --volume-loss-weight 10.0 --vol-loss-delta 1.0 --wandb-group af-huber-volume --wandb-name "tanjiro/af-huber-d1.0" --agent tanjiro --epochs 999
```

### Trial 2 — Huber delta=0.1 (W&B: `33vayqig`)

| Metric | Value | vs Baseline |
|--------|-------|-------------|
| **val surface_mse** | 0.002746 (ep163) | **+827%** ❌ |
| val volume_mse (at best srf) | 0.016576 | +713% ❌ |
| test surface_mse | 0.001796 | — |
| test volume_mse | 0.022078 | — |
| Diverged at | ~ep239 (grad_norm → ∞) | |
| Total epochs | 768 (diverged from ~ep239) | |
| Grad norm (pre-diverge) | 5–7 (over-clamped) | |
| Peak VRAM | ~42 GB | |

**Command:**
```
CUDA_VISIBLE_DEVICES=2 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --ema-decay 0.999 --volume-loss-weight 10.0 --vol-loss-delta 0.1 --wandb-group af-huber-volume --wandb-name "tanjiro/af-huber-d0.1" --agent tanjiro --epochs 999
```

### Trial 3 — Skipped

Condition was both Trial 1 and Trial 2 must improve vol_mse below 0.002039. Neither did, so Trial 3 was not run.

### What happened

1. **delta=1.0 degenerates to MSE.** The researcher agent predicted this: typical per-cell absolute errors are ~0.045, far below delta=1.0. For errors below delta, Huber loss is identical to 0.5×MSE, so this trial effectively ran MSE with a 0.5× scaling factor on the volume loss. The result is a weaker volume gradient signal compared to the baseline MSE, explaining why both surface and volume are worse.

2. **delta=0.1 over-suppresses gradients.** Grad norms were 5–7 (vs 14–22 for delta=1.0 and ~10 for the baseline). The tight Huber clamp limited the model's ability to refine volume predictions, and the surface loss also suffered from the shared backbone receiving weaker volume gradients. The model diverged at ep~239 despite the low grad norms — the clamped gradients may have pushed the optimizer into an unstable region where the surface loss alone couldn't maintain stability.

3. **The fundamental issue:** At this scale of volume errors, the outlier gradient suppression that Huber provides is either irrelevant (delta=1.0, all errors are within the quadratic zone) or counterproductive (delta=0.1, useful gradient signal is destroyed). The baseline MSE + EMA + grad-clip=1.0 already handles gradient stability well — there's no outlier gradient problem to solve.

### Bug fix included

Fixed `primary_metric_key` function shadowing bug on this branch (same fix as PR #3101 — removed duplicate variable assignment block that shadowed the `primary_metric_key()` function with a string variable, causing `UnboundLocalError` on startup).

### Suggested follow-ups

- **Not recommended: intermediate delta values (e.g., 0.3–0.5).** The analysis shows the error scale (~0.045 per-cell) means any delta > ~0.05 degenerates to MSE, and any delta < ~0.05 over-suppresses. There's no productive sweet spot.
- **Alternative robust losses:** If outlier suppression is still a goal, consider **log-cosh loss** (smooth approximation to L1 without a hard transition threshold) or **truncated MSE** (ignore the top-k% of per-cell losses). These avoid the sharp delta threshold that makes Huber fragile.
- **The current champion config (MSE + EMA + vol-weight=10x) may already be near-optimal** for the volume loss function. Volume improvements may come from architectural changes (e.g., volume-specific heads) rather than loss function changes.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3215 because Closing as dead end. W&B confirms both Huber delta trials are significantly worse than the MSE baseline (0.000296 surface_mse):

- **delta=1.0** (run 87hwn7by): best val surface_mse = 0.000484 (+63%) — Huber degenerates to MSE for errors below delta; this run effectively applied a 0.5× scale factor to vol gradients, weakening both surface and volume performance.
- **delta=0.1** (run 33vayqig): best val surface_mse = 0.002746 (+827%), diverged at ep~239 — over-suppressed gradients destabilized optimization.

The student's post-hoc analysis is correct: per-cell errors (~0.045) are already below delta=1.0, so Huber has no outlier-suppression effect at that threshold; any smaller delta destroys useful signal. MSE + EMA + gc=1.0 already handles gradient stability adequately on this dataset.

Hypothesis falsified. No follow-up delta values are warranted — the mechanism doesn't apply here..
